### PR TITLE
Implement On-Chain NFT

### PR DESCRIPTION
**Pull request by shamilsan**
This is an on-chain nft implementation based on the nft implementation in gear-contract-libraries.

**Release Notes**: This PR introduces on-chain NFT smart contract. Instead of storing the content of NFT somewhere else (e.g. IPFS), it is stored directly in the contract on the blockchain. Every time an NFT is queried, instead of the link, a user gets it's content (all the layers). More on that: [wiki](https://wiki.gear-tech.io/examples/onchain-nft).

### CODE REVIEW

File: on-chain-nft/tests/on_chain_tests.rs [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (shamilsan) [inline on-chain-nft/tests/on_chain_tests.rs:1]: Why this file is empty?

File: on-chain-nft/tests/utils.rs [hunk 2]
```diff
@@ -0,0 +1 @@
+
```
Comment (shamilsan) [inline on-chain-nft/tests/utils.rs:1]: Why this file is empty?

File: on-chain-nft/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    let config: InitOnChainNFT = msg::load().expect("Unable to decode InitNFT");
+    let mut _layers: BTreeMap<LayerId, BTreeMap<LayerItemId, Vec<u8>>> = BTreeMap::new();
+    for (layer_id, layer) in config.layers.iter() {
+        let mut layer_map: BTreeMap<LayerItemId, Vec<u8>> = BTreeMap::new();
+        for (layer_item_id, layer_item) in layer.clone() {
+            layer_map.insert(layer_item_id, layer_item.into_bytes());
+        }
+
+        _layers.insert(*layer_id, layer_map);
+    }
```
[the remark this answers is not in the document]
Comment (Akado2009) [reply on-chain-nft/src/lib.rs:34]: Vec<u8> is used for the memory purposes.

File: on-chain-nft/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    let nft = OnChainNFT {
+        token: NFTState {
+            name: config.name,
+            symbol: config.symbol,
+            base_uri: config.base_uri,
+            ..Default::default()
+        },
+        owner: msg::source(),
+        base_image: config.base_image.into_bytes(),
```
[the remark this answers is not in the document]
Comment (Akado2009) [reply on-chain-nft/src/lib.rs:43]: Same goes here.

File: on-chain-nft/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        // check if exists
+        let nft = self.nfts.get(&token_id).expect("No such nft");
+        for (layer_id, layer_item_id) in nft {
+            let layer_content = self
+                .layers
+                .get(layer_id)
+                .expect("No such layer")
+                .get(layer_item_id)
+                .expect("No such layer item");
+            let cc = String::from_utf8((*layer_content).clone()).expect("Found invalid UTF-8");
```
[the remark this answers is not in the document]
Comment (Akado2009) [reply on-chain-nft/src/lib.rs:137]: And finally here.

File: on-chain-nft/io/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,52 @@
+#![no_std]
+
+use codec::{Decode, Encode};
+use gear_lib::non_fungible_token::{royalties::*, token::*};
+use gstd::{prelude::*, ActorId};
+use scale_info::TypeInfo;
+
+pub type LayerId = u128;
+pub type LayerItemId = u128;
```
[the remark this answers is not in the document]
Comment (Akado2009) [reply on-chain-nft/io/src/lib.rs:9]: Nice catch.

File: on-chain-nft/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    fn mint(&mut self, description: BTreeMap<LayerId, LayerItemId>, metadata: TokenMetadata) {
+        // precheck if the layers actually exist
+        for (layer_id, layer_item_id) in description.clone() {
+            let _ = self
+                .layers
+                .get(&layer_id)
+                .expect("No such layer")
+                .get(&layer_item_id)
+                .expect("No such layer item");
+        }
```
[the remark this answers is not in the document]
Comment (Akado2009) [reply on-chain-nft/src/lib.rs:96]: 👍

File: on-chain-nft/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        // precheck if there is already an nft with such description
+        let mut key = String::from("");
+        for lii in description.values() {
+            key = key + &lii.to_string();
+        }
+        if self.nfts_existence.get(&key).is_some() {
+            panic!("Such nft already exists");
+        }
+        self.nfts_existence.insert(key, true);
```
[the remark this answers is not in the document]
Comment (Akado2009) [reply on-chain-nft/src/lib.rs:110]: Yes, great call!!!

File: on-chain-nft/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Akado2009) [reply on-chain-nft/src/lib.rs:43]: Resolved to String, but user will be forced to use a base64encded images.

File: on-chain-nft/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+pub trait OnChainNFTCore: NFTCore {
+    fn mint(&mut self, description: BTreeMap<LayerId, ItemId>, metadata: TokenMetadata);
+    fn burn(&mut self, token_id: TokenId);
+    fn token_uri(&mut self, token_id: TokenId) -> Option<Vec<u8>>;
+}
+
+impl OnChainNFTCore for OnChainNFT {
+    fn mint(&mut self, description: BTreeMap<LayerId, ItemId>, metadata: TokenMetadata) {
+        // precheck if the layers actually exist
+        for (layer_id, layer_item_id) in description.iter() {
```
Comment (LouiseMedova) [inline on-chain-nft/src/lib.rs:89]: `LayerId` and `ItemId` are just numbers, you can just check,  that these numbers lie within [0, layers.len() - 1 ] for `layer_id` and accordingly [0, items.len() - 1 ] for `layer_id`

File: on-chain-nft/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+pub trait OnChainNFTCore: NFTCore {
+    fn mint(&mut self, description: BTreeMap<LayerId, ItemId>, metadata: TokenMetadata);
+    fn burn(&mut self, token_id: TokenId);
+    fn token_uri(&mut self, token_id: TokenId) -> Option<Vec<u8>>;
+}
+
+impl OnChainNFTCore for OnChainNFT {
+    fn mint(&mut self, description: BTreeMap<LayerId, ItemId>, metadata: TokenMetadata) {
```
Comment (LouiseMedova) [inline on-chain-nft/src/lib.rs:87]: if the levels are already stored in a certain order, then maybe it just makes sense to give a variant of these levels in that order? 
```suggestion
    fn mint(&mut self, description: Vec<ItemId>, metadata: TokenMetadata) {
```

Comment (LouiseMedova) [reply on-chain-nft/src/lib.rs:87]: It will also make it easier to check for identical tokens

File: on-chain-nft/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        }
+        self.nfts_existence.insert(key);
+        NFTCore::mint(self, &msg::source(), self.token_id, Some(metadata));
+        self.nfts.insert(self.token_id, description);
+        self.token_id = self.token_id.saturating_add(U256::one());
+    }
+
+    fn burn(&mut self, token_id: TokenId) {
+        NFTCore::burn(self, token_id);
+        self.nfts.remove(&token_id);
```
Comment (LouiseMedova) [inline on-chain-nft/src/lib.rs:118]: You also have to clear the `nft_existence`

File: on-chain-nft/tests/utils.rs [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+use gtest::{Program, RunResult, System};
+use on_chain_nft_io::*;
+const USERS: &[u64] = &[3, 4, 5];
+
+pub fn init_nft(sys: &System) {
+    sys.init_logger();
+    let nft = Program::current(sys);
+
+    let mut layers = BTreeMap::new();
+    let first_layer = BTreeMap::from([
```
Comment (LouiseMedova) [inline on-chain-nft/tests/utils.rs:12]: It must be in `base64` format

File: on-chain-nft/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+#[derive(Debug, Default, NFTStateKeeper, NFTCore, NFTMetaState)]
+pub struct OnChainNFT {
+    #[NFTStateField]
+    pub token: NFTState,
+    pub token_id: TokenId,
+    pub owner: ActorId,
+    pub base_image: String,
+    pub layers: BTreeMap<LayerId, Vec<String>>,
+    pub nfts: BTreeMap<TokenId, Vec<ItemId>>,
```
Comment (LouiseMedova) [inline on-chain-nft/src/lib.rs:17]:
```suggestion
    pub nfts: BTreeMap<TokenId, BTreeSet<ItemId>>,
```

**Comment by shamilsan** [comment]
Stale PR? Need to fix tests and force reviewers to review 😄 
@Akado2009

**Comment by Akado2009** [comment]
> Stale PR? Need to fix tests and force reviewers to review 😄 @Akado2009

OnChain is ready and has been reviewed I believe by @LouiseMedova and @fluiderson. The tests that are crashing are the nft-marketplace tests :( 

As far as I could see the problem is the same in all the tests -- the user performing actions with the auction does not have enough balance. Might try to fix it, but it will take some time.

File: on-chain-nft/io/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+#[derive(Debug, Encode, Decode, TypeInfo)]
+pub enum OnChainNFTAction {
+    /// Mints an NFT.
+    /// Requirements:
+    /// * `description` MUST contain layers and layers' items that EXIST
+    /// Arguments:
+    /// * `token_metadata` - is a default token metadata from gear-lib.
+    /// * `description` - is the vector of layer's item id, where
+    /// the index i is the layer id.
+    /// On success, returns NFTEvent::Mint from gear-lib.
```
Comment (LouiseMedova) [inline on-chain-nft/io/src/lib.rs:31]: More details: mints NFT consisted from layers..
```suggestion
    /// Mints an NFT.
    ///
    /// Requirements:
    /// * `description` MUST contain layers and layers' items that EXIST
    ///
    /// Arguments:
    /// * `token_metadata` - is a default token metadata from gear-lib.
    /// * `description` - is the vector of layer's item id, where
    /// the index i is the layer id.
    ///
    /// On success, returns NFTEvent::Mint from gear-lib.
```

File: on-chain-nft/io/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+        token_metadata: TokenMetadata,
+        /// Layers description of an NFT
+        description: Vec<ItemId>,
+    },
+    /// Burns an NFT.
+    /// Requirements:
+    /// * `token_id` MUST exist
+    /// Arguments:
+    /// * `token_id` - is the id of the burnt token
+    /// On sucess, returns NFTEvent::Burn from gear-lib.
```
Comment (LouiseMedova) [inline on-chain-nft/io/src/lib.rs:43]:
```suggestion
    /// On success, returns NFTEvent::Burn from gear-lib.
```

**Review by Akado2009 — APPROVED** [review]

**Review by LouiseMedova — APPROVED** [review]

**Pull request merged**